keps/sig-architecture/2019-07-16-node-role-label-use.md

```diff
@@ -0,0 +1,215 @@
+  * Deployments should stop setting `LegacyNodeRoleBehavior=off` if they opted out early.
+* Release 4 (1.19):
+  * The `LegacyNodeRoleBehavior` gate and all feature-level gates are removed, components that attempt to set these gates will fail to start.
+  * Code that references node-roles within Kubernetes will be removed.
+
+In Release 4 (which could be as early as 1.19) this KEP will be considered complete.
+
+
+### Current users of `node-role.kubernetes.io/*` within the project that must change
+
+The following components vary behavior based on the presence of the node-role labels:
+
+
+#### Service load-balancer
+
+The service load balancer implementation previously implemented a heuristic where `node-role.kubernetes.io/master` is used to exclude masters
+from the candidate nodes for a service. This is an implementation detail of the cluster and is not allowed. Since there is value in excluding
+nodes from service load balancer candidacy in some deployments, an alpha feature gated label `alpha.service-controller.kubernetes.io/exclude-balancer`
+was added in Kubernetes 1.9.
+
+This label should be moved to beta in Kube 1.16 at its final name `node.kubernetes.io/exclude-balancer`, its feature gate `ServiceNodeExclusion` should default on in 1.17, the gate `ServiceNodeExclusion` should be declared GA in 1.18, and the gate will be removed in 1.19. The old alpha label should be honored in 1.16 and removed in 1.17.
+
+Starting in 1.16 the legacy code block should be gated on `LegacyNodeRoleBehavior=on`
+
+
+#### Node controller excludes master nodes from consideration for eviction
+
+The `k8s.io/kubernetes/pkg/util/system/IsMasterNode(nodeName)` function is used by the NodeLifecycleController to exclude nodes with a node name that ends in `master` or starts with `master-` when considering whether to mark nodes as disrupted. A recent PR attempted to change this to use node-roles and was blocked. Instead, the controller should be updated to use a label `node.kubernetes.io/exclude-disruption` to decide whether to exclude nodes from being considered for disruption handling.
+
+
+#### Kubernetes e2e tests
+
+The e2e tests use a number of heuristics including the `IsMasterNode(nodeName)` function and the node-roles labels to select nodes. In order for conformant Kubernetes clusters to run the tests, the e2e suite must change to use individual user-provided label selectors to identify nodes to test, nodes that have special rules for testing unusual cases, and for other selection behaviors. The label selectors may be defaulted by the test code to their current values, as long as a conformant cluster operator can execute the e2e suite against an arbitrary cluster.
+
+QUESTION: Is a single label selector sufficient to identify nodes to test?
+
+
+#### Preventing accidental reintroduction
```

Comment:
can we clarify how this should work with cloud provider code?  for example, i assume cloud provider (in-tree or out) by convention must not make assumptions on deployment topology?  how can we check for that as well?

Reply:
I added a brief paragraph, but cloud providers are still part of the core K8s rule set because they are owned by a sig and required to maintain conformance on those clouds.

Reply:
how is this distinct from kubeadm exactly?
- owned by a sig
- [maintaining conformance](https://github.com/cncf/k8s-conformance/blob/master/v1.15/kubeadm/)